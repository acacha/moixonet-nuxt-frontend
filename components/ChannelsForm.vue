<template>
  <app-channels-form @added="add" />
</template>

<script>
import * as actionTypes from '../action-types'
import AppChannelsForm from './AppChannelsForm'

export default {
  name: 'ChannelsForm',
  components: {
    'app-channels-form': AppChannelsForm
  },
  methods: {
    add (channel) {
      this.$store.dispatch('channels/' + actionTypes.CHANNEL_STORE, channel)
    }
  }
}
</script>
