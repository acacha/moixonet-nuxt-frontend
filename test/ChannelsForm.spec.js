import { mount } from '@vue/test-utils'
import ChannelsForm from '../components/ChannelsForm'
import channelsFixture from '../cypress/fixtures/channels'

// https://vue-test-utils.vuejs.org/
describe('ChannelsForm', () => {
  let opts

  const build = (options) => {
    options = options || opts
    const wrapper = mount(ChannelsForm, options)
    const channelNameInput = wrapper.find('[data-test=input_name]')
    const addButton = wrapper.find('[data-test=add_button]')

    return {
      wrapper,
      channelNameInput,
      addButton,
      inputErrors: () => wrapper.find('[data-test=input_errors]')
    }
  }

  beforeEach(() => {
    opts = {}
  })

  test('is a Vue instance', () => {
    const { wrapper } = build()
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  test('renders a list of channels', () => {
    const { wrapper } = build()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Can add channel clicking on add button', () => {
    const { wrapper, channelNameInput, addButton, inputErrors } = build()
    channelNameInput.element.value = 'New channel'
    channelNameInput.trigger('input')
    addButton.trigger('click')

    // NO INPUT ERRORS
    expect(inputErrors().exists()).toEqual(false)

    expect(channelNameInput.element.value).toBe('')

    expect(wrapper.emitted().added).toBeTruthy()
    expect(wrapper.emitted().added.length).toBe(1)
    expect(wrapper.emitted().added[0][0]).toEqual('New channel')
  })

  test('Can add channel with enter', () => {
    const { wrapper, channelNameInput } = build()
    channelNameInput.element.value = 'New channel'
    channelNameInput.trigger('input')

    channelNameInput.trigger('keyup.enter')

    expect(channelNameInput.element.value).toBe('')

    expect(wrapper.emitted().added).toBeTruthy()
    expect(wrapper.emitted().added.length).toBe(1)
    expect(wrapper.emitted().added[0][0]).toEqual('New channel')
  })

  test('Show validation error if no channel name provided', () => {
    const { addButton, inputErrors } = build()
    addButton.trigger('click')
    expect(inputErrors().exists()).toEqual(true)
    expect(inputErrors().text()).toEqual('El nom és obligatori')
  })

  test('Validation error is removed after resolving error', () => {
    const { addButton, channelNameInput, inputErrors } = build()
    addButton.trigger('click')
    expect(inputErrors().exists()).toEqual(true)
    expect(inputErrors().text()).toEqual('El nom és obligatori')
    channelNameInput.element.value = 'New channel'
    channelNameInput.trigger('input')
    expect(inputErrors().exists()).toEqual(false)
  })
})
