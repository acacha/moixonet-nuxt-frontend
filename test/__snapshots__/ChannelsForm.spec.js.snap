// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelsForm renders a list of channels 1`] = `
<form>
  <!---->
   
  <input
    data-test="input_name"
    type="text"
  />
   
  <button
    data-test="add_button"
  >
    
    Afegir
  
  </button>
</form>
`;
